File: core/src/nl/t64/game/rpg/screens/world/DebugBox.java
package nl.t64.game.rpg.screens.world;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Pixmap;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.utils.Drawable;
import com.badlogic.gdx.utils.Align;
import nl.t64.game.rpg.Engine;
import nl.t64.game.rpg.components.character.Character;
import nl.t64.game.rpg.constants.Constant;


class DebugBox {

    private static final Color TRANSPARENT = new Color(0f, 0f, 0f, 0.5f);
    private static final float TABLE_WIDTH = 200f;
    private static final float SECOND_COLUMN_WIDTH = 100f;

    private final Stage stage;
    private final Table table;
    private final Character player;


    DebugBox(Character player) {
        this.stage = new Stage();
        this.table = createTable();
        this.player = player;
        this.stage.addActor(this.table);
    }

    void dispose() {
        stage.clear();
        stage.dispose();
    }

    void update(float dt) {
        table.clear();

        table.add("FPS:");
        table.add(String.valueOf(Gdx.graphics.getFramesPerSecond()));
        table.row();
        table.add("dt:");
        table.add(String.valueOf(dt));
        table.row();
        table.add("runTime:");
        String runTime = String.valueOf(Engine.getRunTime());
        table.add(runTime.substring(0, runTime.length() - 4));
        table.row();
        table.add("").row();

//        table.add("timeUp:");
//        table.add(player.getInputComponent().getTimeUp());
//        table.row();
//        table.add("timeDown:");
//        table.add(player.getInputComponent().getTimeDown());
//        table.row();
//        table.add("timeLeft:");
//        table.add(player.getInputComponent().getTimeLeft());
//        table.row();
//        table.add("timeRight:");
//        table.add(player.getInputComponent().getTimeRight());
//        table.row();
//        table.add("timeDelay:");
//        table.add(player.getInputComponent().getTimeDelay());
//        table.row();
//        table.add("").row();
//
//        table.add("oldPositionX:");
//        table.add(String.valueOf(player.getOldPosition().x));
//        table.row();
//        table.add("oldPositionY:");
//        table.add(String.valueOf(player.getOldPosition().y));
//        table.row();
//        table.add("").row();

        table.add("currentPositionX:");
        table.add(String.valueOf(player.getPosition().x));
        table.row();
        table.add("currentPositionY:");
        table.add(String.valueOf(player.getPosition().y));
        table.row();
        table.add("").row();

        table.add("currentPositionGridX:");
        table.add(String.valueOf(player.getPositionInGrid().x));
        table.row();
        table.add("currentPositionGridY:");
        table.add(String.valueOf(player.getPositionInGrid().y));
        table.row();
        table.add("").row();

        table.add("currentPositionTiledX:");
        table.add(String.valueOf((int) (player.getPosition().x / Constant.TILE_SIZE)));
        table.row();
        table.add("currentPositionTiledY:");
        table.add(String.valueOf((int) (player.getPosition().y / Constant.TILE_SIZE)));
        table.row();
        table.add("").row();

        table.add("direction:");
        table.add(String.valueOf(player.getDirection()));
        table.row();
        table.add("state:");
        table.add(String.valueOf(player.getState()));
        table.row();
        table.add("").row();

        table.pack();
        table.setPosition(0, Gdx.graphics.getHeight() - table.getHeight());

        stage.act(dt);
        stage.draw();
    }

    private Table createTable() {
        var debugSkin = new Skin();
        debugSkin.add("default", new Label.LabelStyle(new BitmapFont(), Color.WHITE));

        var newTable = new Table(debugSkin);
        newTable.defaults().width(TABLE_WIDTH).align(Align.left);
        newTable.columnDefaults(1).width(SECOND_COLUMN_WIDTH);
        newTable.setBackground(createTransparency());

        return newTable;
    }

    private Drawable createTransparency() {
        var pixmap = new Pixmap(1, 1, Pixmap.Format.Alpha);
        pixmap.setColor(TRANSPARENT);
        pixmap.fill();
        return new Image(new Texture(pixmap)).getDrawable();
    }

}

File: core/src/nl/t64/game/rpg/screens/world/DebugBox.kt
package nl.t64.game.rpg.screens.world

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.Pixmap
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.utils.Drawable
import com.badlogic.gdx.utils.Align
import nl.t64.game.rpg.Engine
import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.constants.Constant
import nl.t64.game.rpg.screens.world.entity.Entity


private val TRANSPARENT = Color(0f, 0f, 0f, 0.5f)
private const val TABLE_WIDTH = 200f
private const val SECOND_COLUMN_WIDTH = 100f

internal class DebugBox(private val player: Entity) {

    private val table: Table = createTable()
    private val stage: Stage = Stage().apply { addActor(table) }

    fun dispose() {
        stage.dispose()
    }

    fun update(dt: Float) {
        table.clear()

        table.add("FPS:")
        table.add(Gdx.graphics.framesPerSecond.toString())
        table.row()
        table.add("dt:")
        table.add(dt.toString())
        table.row()
        table.add("runTime:")
        val runTime = Engine.getRunTime().toString()
        table.add(runTime.substring(0, runTime.length - 4))
        table.row()
        table.add("").row()

//        table.add("timeUp:");
//        table.add(player.getInputComponent().getTimeUp());
//        table.row();
//        table.add("timeDown:");
//        table.add(player.getInputComponent().getTimeDown());
//        table.row();
//        table.add("timeLeft:");
//        table.add(player.getInputComponent().getTimeLeft());
//        table.row();
//        table.add("timeRight:");
//        table.add(player.getInputComponent().getTimeRight());
//        table.row();
//        table.add("turnDelay:");
//        table.add(String.valueOf(player.getPlayerInput().getTurnDelay()));
//        table.row();
//        table.add("turnGrace:");
//        table.add(String.valueOf(player.getPlayerInput().getTurnGrace()));
//        table.row();
//        table.add("").row();
//
//        table.add("oldPositionX:");
//        table.add(String.valueOf(player.getOldPosition().x));
//        table.row();
//        table.add("oldPositionY:");
//        table.add(String.valueOf(player.getOldPosition().y));
//        table.row();
//        table.add("").row();

        table.add("currentPositionX:")
        table.add(player.position.x.toString())
        table.row()
        table.add("currentPositionY:")
        table.add(player.position.y.toString())
        table.row()
        table.add("").row()

        table.add("currentPositionGridX:")
        table.add(player.getPositionInGrid().x.toString())
        table.row()
        table.add("currentPositionGridY:")
        table.add(player.getPositionInGrid().y.toString())
        table.row()
        table.add("").row()

        table.add("currentPositionTiledX:")
        table.add((player.position.x / Constant.TILE_SIZE).toInt().toString())
        table.row()
        table.add("currentPositionTiledY:")
        table.add((player.position.y / Constant.TILE_SIZE).toInt().toString())
        table.row()
        table.add("").row()

        table.add("direction:")
        table.add(player.direction.toString())
        table.row()
        table.add("state:")
        table.add(player.state.toString())
        table.row()
        table.add("").row()

        table.add("quest0001:")
        table.add(gameData.quests.getQuestById("quest0001").currentState.toString())
        table.add("").row()
        table.add("quest0002:")
        table.add(gameData.quests.getQuestById("quest0002").currentState.toString())
        table.add("").row()
        table.add("quest0003:")
        table.add(gameData.quests.getQuestById("quest0003").currentState.toString())
        table.add("").row()
        table.add("quest0004:")
        table.add(gameData.quests.getQuestById("quest0004").currentState.toString())
        table.add("").row()
        table.add("quest0005:")
        table.add(gameData.quests.getQuestById("quest0005").currentState.toString())
        table.add("").row()

        table.pack()
        table.setPosition(0f, Gdx.graphics.height - table.height)

        stage.act(dt)
        stage.draw()
    }

    private fun createTable(): Table {
        val debugSkin = Skin()
        debugSkin.add("default", LabelStyle(BitmapFont(), Color.WHITE))

        return Table(debugSkin).apply {
            defaults().width(TABLE_WIDTH).align(Align.left)
            columnDefaults(1).width(SECOND_COLUMN_WIDTH)
            background = createTransparency()
        }
    }

    private fun createTransparency(): Drawable {
        val pixmap = Pixmap(1, 1, Pixmap.Format.Alpha)
        pixmap.setColor(TRANSPARENT)
        pixmap.fill()
        return Image(Texture(pixmap)).drawable
    }

}
